cns/fsnotify/fsnotify.go

```diff
@@ -139,51 +160,19 @@ func (w *watcher) watchFS(ctx context.Context) error {
 	}
 }
 
-// readFS lists the directory and enqueues any missed deletes that are already
-// present on-disk.
-func (w *watcher) readFS() error {
-	w.log.Info("listing directory", zap.String("path", w.path))
-	dirContents, err := os.ReadDir(w.path)
-	if err != nil {
-		w.log.Error("error reading deleteID directory", zap.String("path", w.path), zap.Error(err))
-		return errors.Wrapf(err, "failed to read %s", w.path)
-	}
-	if len(dirContents) == 0 {
-		w.log.Info("no missed deletes found")
-		return nil
-	}
-	w.lock.Lock()
-	for _, file := range dirContents {
-		w.log.Info("adding missed delete from file", zap.String("name", file.Name()))
-		w.pendingDelete[file.Name()] = struct{}{}
-	}
-	w.lock.Unlock()
-	return nil
-}
-
-// WatchFS starts the filesystem watcher to handle async Pod deletes.
+// Start starts the filesystem watcher to handle async Pod deletes.
 // Blocks until the context is closed; returns underlying fsnotify errors
 // if something goes fatally wrong.
-func (w *watcher) Start(ctx context.Context) error {
-	errs := make(chan error)
+func (w *watcher) Start(c context.Context) error {
+	ctx, cancel := context.WithCancel(c)
+	defer cancel()
+	g, groupCtx := errgroup.WithContext(ctx)
 	// Start watching for enqueued missed deletes so that we process them as soon as they arrive.
-	go func(errs chan<- error) {
-		errs <- w.watchPendingDelete(ctx)
-	}(errs)
-
+	g.Go(func() error { return w.watchPendingDelete(groupCtx) })
 	// Start watching for changes to the filesystem so that we don't miss any async deletes.
-	go func(errs chan<- error) {
-		errs <- w.watchFS(ctx)
-	}(errs)
-
-	// Read the directory to enqueue any missed deletes that are already present on-disk.
-	if err := w.readFS(); err != nil {
-		return err
-	}
-
+	g.Go(func() error { return w.watchFS(groupCtx) })
 	// block until one of the goroutines returns an error
```

Comment:
nit: this comment might now be better with the errgroup semantics:
```
// the first error from the errgroup will trigger context cancellation for other goroutines in the group.
// this will block until all goroutines complete and return the first error.
```

Reply:
updated